### PR TITLE
feat: display control modification icon by position

### PR DESCRIPTION
**Pull request by mpemy**
Closes: #82

### CODE REVIEW

File: packages/oscal-react-library/src/components/OSCALControl.tsx [hunk 1]
```diff
@@ -241,9 +241,139 @@ const OSCALControl: React.FC<OSCALControlProps> = (props) => {
...
 
+  const position = props?.modificationAlters
+    ?.find((item) => item["control-id"] === props.control.id)
+    ?.adds?.flatMap((item) => item["position"])
+    .filter((part) => part)[0];
+
   const label = propWithName(control.props, "label")?.value;
   const controlOrParentWithdrawn = withdrawn || isWithdrawn(control);
 
+  if (position === "after") {
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALControl.tsx:252]: Can you add `import { Position } from "@easydynamics/oscal-types";` to the top of this file and then use the enum variants in these conditions (`if (position === Position.AFTER) {`)?

I do think we need to find a slightly less-repetitive way to handle the code within the conditions but I am not quite sure how to go about that refactoring. I think @tuckerzp is pretty familiar with this so maybe he can provide an opinion

Comment (mpemy) [reply packages/oscal-react-library/src/components/OSCALControl.tsx:252]: Thanks @kylelaker for this suggestion, I will do that.

Comment (tuckerzp) [reply packages/oscal-react-library/src/components/OSCALControl.tsx:252]: > a slightly less-repetitive way to handle the code within the conditions

Briefly looking at this, the only difference I see here for each if is the position of `modifcationDisplay` in the component being created.

At the very least, could we not do something like:

```TypeScript
<OSCALControlCard>
  <position1>
    {position === Position.AFTER) ? modificationDisplay : null}
  </position1>
  <position2>
    {position === Position.Before) ? modificationDisplay : null}
  </position2>
</OSCALControlCard>
```

I don't think we need to create a completely different declaration of the component if we are just switching 1 thing around based on an if statement.

Comment (mpemy) [reply packages/oscal-react-library/src/components/OSCALControl.tsx]: Thanks @tuckerzp, I refactored the code. 
This is the profile I am using to test the implementation. https://raw.githubusercontent.com/EasyDynamics/oscal-demo-content/test-content/profiles/NIST_SP-800_rev4TestProfile.json
These are few screenshots
image

image

File: packages/oscal-react-library/src/components/OSCALControlPart.js [hunk 2]
```diff
@@ -44,35 +50,73 @@ export default function OSCALControlPart(props) {
...
-        componentId={props.componentId}
-        modificationSetParameters={props.modificationSetParameters}
-        modificationDisplay={modificationDisplay}
-        isEditable={props.isEditable}
-        onRestSuccess={props.onRestSuccess}
-        onRestError={props.onRestError}
-        partialRestData={props.partialRestData}
-      />
+      <Typography>
+        {modificationDisplayBefore}
```
[the remark this answers is not in the document]
Comment (mpemy) [reply packages/oscal-react-library/src/components/OSCALControlPart.js:88]: @tuckerzp I tried that approach first, but you can see that when I defined modificationDisplayBefore, I have to turn off all the other modificationDisplay so I have more than one instruction to be executed, that is why I set all other modificationDisplay to undefined. I first try to set them to null, but I got some linting issues.

**Comment by tuckerzp** [comment]
Looking at the screenshot now, I also am wondering about the position of the modification icon. Why are they now on a completely new line? 

For example, `ac-2_smt.a` has modifications on the line above. I wonder if it would look better like.

```
a. {icon} identifies...
````

It may be helpful to describe where you think modifications should appear for each `position` value. It can even be a  quick code block like the one above.

File: packages/oscal-react-library/src/components/OSCALControl.tsx [hunk 3]
```diff
@@ -241,9 +242,50 @@ const OSCALControl: React.FC<OSCALControlProps> = (props) => {
...
+
+  switch (position) {
+    case Position.AFTER:
+      // eslint-disable-next-line no-lone-blocks
+      {
+        modificationDisplayBefore = undefined;
+        modificationDisplayEnding = undefined;
+        modificationDisplayStarting = undefined;
+      }
+      break;
```
[the remark this answers is not in the document]
Comment (mpemy) [reply packages/oscal-react-library/src/components/OSCALControl.tsx:266]: I kept the switch block and added the break clause within the curly brackets.

File: packages/oscal-react-library/src/components/OSCALControl.tsx [hunk 4]
```diff
@@ -241,9 +242,44 @@ const OSCALControl: React.FC<OSCALControlProps> = (props) => {
...
+    ?.adds?.flatMap((item) => item["position"])
+    .filter((part) => part)[0];
+
   const label = propWithName(control.props, "label")?.value;
   const controlOrParentWithdrawn = withdrawn || isWithdrawn(control);
 
+  let modificationDisplayBefore = modificationDisplay;
+  let modificationDisplayAfter = modificationDisplay;
+  let modificationDisplayEnding = modificationDisplay;
+  let modificationDisplayStarting = modificationDisplay;
```
Comment (tuckerzp) [inline packages/oscal-react-library/src/components/OSCALControl.tsx:256]: > but you can see that when I defined modificationDisplayBefore, I have to turn off all the other modificationDisplay

```suggestion
  const modificationDisplayBefore = position === Position.BEFORE ? modificationDisplay : null;
  const modificationDisplayAfter = position === Position.AFTER ? modificationDisplay : null;;
  const modificationDisplayEnding = position === Position.ENDING ? modificationDisplay : null;;
  const modificationDisplayStarting = position === Position.STARTING ? modificationDisplay : null;;
```

I am pretty sure this is equivelent to what you are doing right? If `position == BEFORE`, all other the other variables will be `null`. You also could just do this check everywhere you put these variables, but that is up to personal preference. 

This makes it so we do not have to use `let` and it makes the code a lot clearer. I really think there must be a cleaner way to deal with this other than the `switch`.

File: packages/oscal-react-library/src/components/OSCALControlPart.js [hunk 5]
```diff
@@ -117,7 +155,31 @@ export default function OSCALControlPart(props) {
...
+            modificationAlters={props.modificationAlters}
+            modificationSetParameters={props.modificationSetParameters}
+            key={part.id ?? part.name}
+            isEditable={props.isEditable}
+            onRestSuccess={props.onRestSuccess}
+            onRestError={props.onRestError}
+            partialRestData={props.partialRestData}
+          />
+        ))}
+        {modificationDisplay}
```
Comment (tuckerzp) [inline packages/oscal-react-library/src/components/OSCALControlPart.js:179]: Why is there not a `modificationDisplayEnding` in this file?

**Review by tuckerzp — CHANGES_REQUESTED** [review]

File: packages/oscal-react-library/src/components/OSCALControlPart.js (hunk 5 above)
Comment (mpemy) [reply packages/oscal-react-library/src/components/OSCALControlPart.js]: This block of codes in under the condition that position == Position.ENDING, that is why at first, I did not use modificationDisplayEnding there. But for code clarity, I just replaced modificationDisplay by modificationDisplayEnding.
Thanks!

**Comment by tuckerzp** [comment]
I made a (hopefully) simpler emaple Profile to test with. Feel free to make any changes you feel you'd want.

https://raw.githubusercontent.com/EasyDynamics/oscal-demo-content/tuckerzp/profile-mods/profiles/test-profile.json

image

---

@mpemy is this loading like intended? It is still a bit unclear to me where you intend modifications to be for each position.

**Comment by mpemy** [comment]
@tuckerzp I have addressed the positioning issue. The implementation follows acceptance criteria: 
before: Preceding the id-ref target
after: Following the id-ref target
starting: Inside the control or id-ref target, at the start
ending: Inside the control or id-ref target, at the end.

Thanks!

**Comment by mpemy** [comment]
@tuckerzp  This is a screenshot using your profile.
image

**Review by kylelaker — APPROVED** [review]
Thanks for your work on this and for collaborating with @tuckerzp. I think this is a clean implementation that meets the ticket. We might need to more largely re-evaluate how modifications are handled in the future.

**Comment by mpemy** [comment]
You are welcome @kylelaker.

**Review by tuckerzp — APPROVED** [review]

**Pull request merged**